### PR TITLE
feat(server): Move entity BI events to server

### PR DESCRIPTION
**Pull request by yuval-hazaz**
Close: https://github.com/amplication/private-issues/issues/3

## PR Details

move entity events from client to server side 

## PR Checklist

- [ ] Tests for the changes have been added
- [ ] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### Summary
🗑️🚀🧪

Removed client-side analytics tracking for entity and entity field operations and added server-side analytics tracking using `SegmentAnalyticsService` and `EnumEventType`. Updated tests and enums accordingly.

> _No more client-side tracking, we send it to the server_
> _We use the SegmentAnalyticsService to make them shiver_
> _We define the EnumEventType for every entity action_
> _We test the analytics logic with satisfaction_

### Walkthrough
*  Remove unused imports and calls of `useTracking` and `AnalyticsEventNames` from client files ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a268f028ea1e884aa67d49c9e9e7ead860be1bcb6d28faf31da95fdd5a73c96eL14-R14), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a268f028ea1e884aa67d49c9e9e7ead860be1bcb6d28faf31da95fdd5a73c96eL20), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a54127751135be6c11a4a7e2da3089ecdf8dab44438e76e96c60c72a5e77a1f4L10), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a54127751135be6c11a4a7e2da3089ecdf8dab44438e76e96c60c72a5e77a1f4L20), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-27094a723fd0057b3bfae629d22fc14909c04783ff7d3d566cf48fd04f768f65L16-L17), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-f0b6cc1f93832f4e7b7040eb02a5f09079a3fba05d7e07bd1d8819e1a22baaedL9-L10))
*  Remove unused variables and calls of `trackEvent` from client components ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a268f028ea1e884aa67d49c9e9e7ead860be1bcb6d28faf31da95fdd5a73c96eL36), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a268f028ea1e884aa67d49c9e9e7ead860be1bcb6d28faf31da95fdd5a73c96eL50-L53), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a54127751135be6c11a4a7e2da3089ecdf8dab44438e76e96c60c72a5e77a1f4L33), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a54127751135be6c11a4a7e2da3089ecdf8dab44438e76e96c60c72a5e77a1f4L89-L93), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-27094a723fd0057b3bfae629d22fc14909c04783ff7d3d566cf48fd04f768f65L57), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-27094a723fd0057b3bfae629d22fc14909c04783ff7d3d566cf48fd04f768f65L67-R64), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-f0b6cc1f93832f4e7b7040eb02a5f09079a3fba05d7e07bd1d8819e1a22baaedL30), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-f0b6cc1f93832f4e7b7040eb02a5f09079a3fba05d7e07bd1d8819e1a22baaedL74-L78))
*  Remove unused enum values from `AnalyticsEventNames` in client util file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-917333793bfc12bbce1fed70cf28308c74cf550fb42d3653f076a65ad209e697L62-R62))
*  Add `Account` import and constants to server test file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216R20), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216R271-R287))
*  Add `project` and `account` properties to test constants in server test file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216R161-R167), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216R293))
*  Add `prismaResourceFindUniqueMock` function to server test file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216R301-R304))
*  Add `SegmentAnalyticsService` and `prismaResourceFindUniqueMock` providers to server test module ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216L388-R431))
*  Add `EnumEventType` and `SegmentAnalyticsService` imports to server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R79-R82))
*  Add `analytics` parameter to server service constructor ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R179))
*  Add `resourceWithProject` variable and `analytics.track` call to `createEntity` method in server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R330-R350))
*  Add `resourceWithProject` variable and `analytics.track` call to `updateEntity` method in server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R646-R665))
*  Add `resourceWithProject` variable and `analytics.track` call to `createEntityField` method in server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R1976-R1996))
*  Add `resourceWithProject` variable and `analytics.track` call to `updateEntityField` method in server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R2305-R2326))
*  Add event types for entity creation, update, field creation, and field update to `EnumEventType` in server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-18685efac1e573e80e4297b774d6d17dae8ccbbcde9db01faa86aa7740526745R15-R18))

 (for the complete comment)

### <samp>🤖 Generated by Copilot at 78f5e0e</samp>

Removed client-side analytics tracking for entity and entity field operations and added server-side analytics tracking using `SegmentAnalyticsService` and `EnumEventType`. Updated tests and enums accordingly.

### <samp>🤖 Generated by Copilot at 78f5e0e</samp>

*  Remove unused imports and calls of `useTracking` and `AnalyticsEventNames` from client files ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a268f028ea1e884aa67d49c9e9e7ead860be1bcb6d28faf31da95fdd5a73c96eL14-R14), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a268f028ea1e884aa67d49c9e9e7ead860be1bcb6d28faf31da95fdd5a73c96eL20), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a54127751135be6c11a4a7e2da3089ecdf8dab44438e76e96c60c72a5e77a1f4L10), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a54127751135be6c11a4a7e2da3089ecdf8dab44438e76e96c60c72a5e77a1f4L20), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-27094a723fd0057b3bfae629d22fc14909c04783ff7d3d566cf48fd04f768f65L16-L17), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-f0b6cc1f93832f4e7b7040eb02a5f09079a3fba05d7e07bd1d8819e1a22baaedL9-L10))
*  Remove unused variables and calls of `trackEvent` from client components ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a268f028ea1e884aa67d49c9e9e7ead860be1bcb6d28faf31da95fdd5a73c96eL36), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a268f028ea1e884aa67d49c9e9e7ead860be1bcb6d28faf31da95fdd5a73c96eL50-L53), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a54127751135be6c11a4a7e2da3089ecdf8dab44438e76e96c60c72a5e77a1f4L33), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-a54127751135be6c11a4a7e2da3089ecdf8dab44438e76e96c60c72a5e77a1f4L89-L93), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-27094a723fd0057b3bfae629d22fc14909c04783ff7d3d566cf48fd04f768f65L57), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-27094a723fd0057b3bfae629d22fc14909c04783ff7d3d566cf48fd04f768f65L67-R64), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-f0b6cc1f93832f4e7b7040eb02a5f09079a3fba05d7e07bd1d8819e1a22baaedL30), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-f0b6cc1f93832f4e7b7040eb02a5f09079a3fba05d7e07bd1d8819e1a22baaedL74-L78))
*  Remove unused enum values from `AnalyticsEventNames` in client util file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-917333793bfc12bbce1fed70cf28308c74cf550fb42d3653f076a65ad209e697L62-R62))
*  Add `Account` import and constants to server test file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216R20), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216R271-R287))
*  Add `project` and `account` properties to test constants in server test file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216R161-R167), [link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216R293))
*  Add `prismaResourceFindUniqueMock` function to server test file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216R301-R304))
*  Add `SegmentAnalyticsService` and `prismaResourceFindUniqueMock` providers to server test module ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-aa07326c891b090e3d5c18991a6e07f30aaa0b8e581e6ffc9903e8c23ab5d216L388-R431))
*  Add `EnumEventType` and `SegmentAnalyticsService` imports to server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R79-R82))
*  Add `analytics` parameter to server service constructor ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R179))
*  Add `resourceWithProject` variable and `analytics.track` call to `createEntity` method in server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R330-R350))
*  Add `resourceWithProject` variable and `analytics.track` call to `updateEntity` method in server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R646-R665))
*  Add `resourceWithProject` variable and `analytics.track` call to `createEntityField` method in server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R1976-R1996))
*  Add `resourceWithProject` variable and `analytics.track` call to `updateEntityField` method in server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-9a1aca769e1c8ddba4436f7b850d36912e8fa38e946d5c40d68631be40167e59R2305-R2326))
*  Add event types for entity creation, update, field creation, and field update to `EnumEventType` in server service file ([link](https://github.com/amplication/amplication/pull/6022/files?diff=unified&w=0#diff-18685efac1e573e80e4297b774d6d17dae8ccbbcde9db01faa86aa7740526745R15-R18))

### <samp>🤖 Generated by Copilot at 78f5e0e</samp>

> _No more client-side tracking, we send it to the server_
> _We use the SegmentAnalyticsService to make them shiver_
> _We define the EnumEventType for every entity action_
> _We test the analytics logic with satisfaction_

### CODE REVIEW

File: packages/amplication-server/src/core/entity/entity.service.ts [hunk 1]
```diff
@@ -617,6 +643,26 @@ export class EntityService {
...
       }
 
+      const resourceWithProject = await this.prisma.resource.findUnique({
+        where: {
+          id: entity.resourceId,
+        },
+        include: {
+          project: true,
+        },
+      });
```
Comment (abrl91) [inline packages/amplication-server/src/core/entity/entity.service.ts:653]: I noticed we use `resourceWithProject` in more than one place.
We can add a method to the class

```ts
async resourceWithProject(id: string) {
this.prisma.resource.findUnique({
        where: {
          id,
        },
        include: {
          project: true,
        },
      });}
```

Comment (abrl91) [reply packages/amplication-server/src/core/entity/entity.service.ts:653]: and call it with the args id where needed

Comment (yuval-hazaz) [reply packages/amplication-server/src/core/entity/entity.service.ts:653]: Thanks for the suggestion, I was thinking about it but I don't think it is really beneficial and decided not to

If we are looking to keep things DRY I was thinking about adding the method into LockEntity in the finally clause - but decided to keep it is

File: packages/amplication-server/src/core/entity/entity.service.ts [hunk 2]
```diff
@@ -322,6 +327,27 @@ export class EntityService {
         },
       },
     });
+
+    const resourceWithProject = await this.prisma.resource.findUnique({
```
[the remark this answers is not in the document]
Comment (yuval-hazaz) [reply packages/amplication-server/src/core/entity/entity.service.ts:331]: Assuming we had different repositories for different entities (in a complete repository pattern), we could inject multiple repositories to a single service and would not go for a specific service to use a specific repository. 

If you want to implement that, I don't mind, but I will not have time soon to spend on that.

**Review by abrl91 — APPROVED** [review]

**Pull request merged**